pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+    :math:`f: \mathbb{R}^N \rightarrow \mathbb{R}` (as the observable is Hermitian,
+    the expectation value is real-valued).
+    This function can always be expressed by a Fourier-type sum
+
+    .. math::
+
+        \sum \limits_{\omega_1\in \Omega_1} \dots \sum \limits_{\omega_N \in \Omega_N}
+        c_{\omega_1,\dots, \omega_N} e^{-i x_1 \omega_1} \dots e^{-i x_N \omega_N}
+
+    over the *frequency spectra* :math:`\Omega_j \subseteq \mathbb{R},`
+    :math:`j=1,\dots,N`. Each spectrum has the property that
+    :math:`0 \in \Omega_j`, and the spectrum is symmetric
+    (i.e., for every :math:`\omega \in \Omega_j` we have that :math:`-\omega \in\Omega_j`).
+    If all frequencies are integer-valued, the Fourier sum becomes a *Fourier series*.
+
+    As shown in `Vidal and Theis (2019) <https://arxiv.org/abs/1901.11434>`_ and
+    `Schuld, Sweke and Meyer (2020) <https://arxiv.org/abs/2008.08605>`_,
+    if an input :math:`x_j, j = 1 \dots N`,
+    only enters into single-parameter gates of the form :math:`e^{-i x_j G}`
+    (where :math:`G` is a Hermitian generator),
+    the frequency spectrum :math:`\Omega_j` is fully determined by the eigenvalues
+    of the generators :math:`G`. In many situations, the spectra are limited
+    to a few frequencies only, which in turn limits the function class that the circuit
+    can express.
+
+    The ``advanced_spectrum`` function computes all frequencies that will
+    potentially appear in the sets :math:`\Omega_1` to :math:`\Omega_N`.
+
+    .. note::
+
+        In more detail, the ``advanced_spectrum`` function also allows for
+        preprocessing of the QNode arguments before they are fed into the gates,
+        as long as this processing is *linear*. In particular, constant
+        prefactors for the encoding arguments are allowed.
+
+    **Example**
+
+    Consider the following example, which uses non-trainable inputs ``x``, ``y`` and ``z``
+    as well as trainable parameters ``w`` as arguments to the QNode.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from advanced_spectrum import advanced_spectrum
+
+        n_qubits = 3
+        dev = qml.device("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev)
+        def circuit(x, y, z, w):
+            for i in range(n_qubits):
+                qml.RX(0.5*x[i], wires=i)
+                qml.Rot(w[0,i,0], w[0,i,1], w[0,i,2], wires=i)
+                qml.RY(2.3*y[i], wires=i)
+                qml.Rot(w[1,i,0], w[1,i,1], w[1,i,2], wires=i)
+                qml.RX(z, wires=i)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = np.array([1., 2., 3.])
+        y = np.array([0.1, 0.3, 0.5])
+        z = -1.8
+        w = np.random.random((2, n_qubits, 3))
+        res = advanced_spectrum(circuit, argnum=[0, 1, 2])(x, y, z, w)
+
+    This circuit looks as follows:
+
+    >>> print(qml.draw(circuit)(x, y, z, w))
```

Comment:
```suggestion
    This circuit looks as follows:
    
    >>> x = np.array([1., 2., 3.])
    >>> y = np.array([0.1, 0.3, 0.5])
    >>> z = -1.8
    >>> w = np.random.random((2, n_qubits, 3))
    >>> print(qml.draw(circuit)(x, y, z, w))
```

Reply:
I think better to use the spectrum function execution down where it is mentioned!